File: go/internal/service/relationship_service.go
package service

import (
	"context"
	"fmt"

	"github.com/hemanthpathath/flex-db/go/internal/repository"
)

// RelationshipService handles relationship business logic
type RelationshipService struct {
	repo     repository.RelationshipRepository
	nodeRepo repository.NodeRepository
}

// NewRelationshipService creates a new RelationshipService
func NewRelationshipService(repo repository.RelationshipRepository, nodeRepo repository.NodeRepository) *RelationshipService {
	return &RelationshipService{repo: repo, nodeRepo: nodeRepo}
}

// Create creates a new relationship
func (s *RelationshipService) Create(ctx context.Context, tenantID, sourceNodeID, targetNodeID, relType, data string) (*repository.Relationship, error) {
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}
	if sourceNodeID == "" {
		return nil, fmt.Errorf("source_node_id is required")
	}
	if targetNodeID == "" {
		return nil, fmt.Errorf("target_node_id is required")
	}
	if relType == "" {
		return nil, fmt.Errorf("relationship_type is required")
	}

	// Validate that the source node belongs to the same tenant
	sourceNode, err := s.nodeRepo.GetByID(ctx, tenantID, sourceNodeID)
	if err != nil {
		return nil, fmt.Errorf("invalid source_node_id: node not found or does not belong to this tenant")
	}
	if sourceNode.TenantID != tenantID {
		return nil, fmt.Errorf("invalid source_node_id: node does not belong to this tenant")
	}

	// Validate that the target node belongs to the same tenant
	targetNode, err := s.nodeRepo.GetByID(ctx, tenantID, targetNodeID)
	if err != nil {
		return nil, fmt.Errorf("invalid target_node_id: node not found or does not belong to this tenant")
	}
	if targetNode.TenantID != tenantID {
		return nil, fmt.Errorf("invalid target_node_id: node does not belong to this tenant")
	}

	rel := &repository.Relationship{
		TenantID:         tenantID,
		SourceNodeID:     sourceNodeID,
		TargetNodeID:     targetNodeID,
		RelationshipType: relType,
		Data:             data,
	}

	return s.repo.Create(ctx, rel)
}

// GetByID retrieves a relationship by ID
func (s *RelationshipService) GetByID(ctx context.Context, tenantID, id string) (*repository.Relationship, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}
	return s.repo.GetByID(ctx, tenantID, id)
}

// Update updates an existing relationship
func (s *RelationshipService) Update(ctx context.Context, tenantID, id, relType, data string) (*repository.Relationship, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}

	rel, err := s.repo.GetByID(ctx, tenantID, id)
	if err != nil {
		return nil, err
	}

	if relType != "" {
		rel.RelationshipType = relType
	}
	if data != "" {
		rel.Data = data
	}

	return s.repo.Update(ctx, rel)
}

// Delete deletes a relationship
func (s *RelationshipService) Delete(ctx context.Context, tenantID, id string) error {
	if id == "" {
		return fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return fmt.Errorf("tenant_id is required")
	}
	return s.repo.Delete(ctx, tenantID, id)
}

// List retrieves relationships with pagination and optional filtering
func (s *RelationshipService) List(ctx context.Context, tenantID, sourceNodeID, targetNodeID, relType string, pageSize int32, pageToken string) ([]*repository.Relationship, *repository.ListResult, error) {
	if tenantID == "" {
		return nil, nil, fmt.Errorf("tenant_id is required")
	}

	opts := repository.ListOptions{
		PageSize:  int(pageSize),
		PageToken: pageToken,
	}
	return s.repo.List(ctx, tenantID, sourceNodeID, targetNodeID, relType, opts)
}

File: python/app/service/relationship_service.py
"""
Relationship service implementation.
"""

from typing import List, Optional, Tuple

from app.repository import Relationship, RelationshipRepository, NodeRepository, ListOptions, ListResult


class RelationshipService:
    """Relationship business logic service."""

    def __init__(self, repo: RelationshipRepository, node_repo: NodeRepository):
        self.repo = repo
        self.node_repo = node_repo

    async def create(
        self,
        tenant_id: str,
        source_node_id: str,
        target_node_id: str,
        rel_type: str,
        data: str
    ) -> Relationship:
        """Create a new relationship."""
        if not tenant_id:
            raise ValueError("tenant_id is required")
        if not source_node_id:
            raise ValueError("source_node_id is required")
        if not target_node_id:
            raise ValueError("target_node_id is required")
        if not rel_type:
            raise ValueError("relationship_type is required")

        # Validate that the source node belongs to the same tenant
        source_node = await self.node_repo.get_by_id(tenant_id, source_node_id)
        if source_node.tenant_id != tenant_id:
            raise ValueError("invalid source_node_id: node does not belong to this tenant")

        # Validate that the target node belongs to the same tenant
        target_node = await self.node_repo.get_by_id(tenant_id, target_node_id)
        if target_node.tenant_id != tenant_id:
            raise ValueError("invalid target_node_id: node does not belong to this tenant")

        rel = Relationship(
            tenant_id=tenant_id,
            source_node_id=source_node_id,
            target_node_id=target_node_id,
            relationship_type=rel_type,
            data=data,
        )
        return await self.repo.create(rel)

    async def get_by_id(self, tenant_id: str, id: str) -> Relationship:
        """Retrieve a relationship by ID."""
        if not id:
            raise ValueError("id is required")
        if not tenant_id:
            raise ValueError("tenant_id is required")
        return await self.repo.get_by_id(tenant_id, id)

    async def update(self, tenant_id: str, id: str, rel_type: str, data: str) -> Relationship:
        """Update an existing relationship."""
        if not id:
            raise ValueError("id is required")
        if not tenant_id:
            raise ValueError("tenant_id is required")

        rel = await self.repo.get_by_id(tenant_id, id)

        if rel_type:
            rel.relationship_type = rel_type
        if data:
            rel.data = data

        return await self.repo.update(rel)

    async def delete(self, tenant_id: str, id: str) -> None:
        """Delete a relationship."""
        if not id:
            raise ValueError("id is required")
        if not tenant_id:
            raise ValueError("tenant_id is required")
        await self.repo.delete(tenant_id, id)

    async def list(
        self,
        tenant_id: str,
        source_node_id: Optional[str],
        target_node_id: Optional[str],
        rel_type: Optional[str],
        page_size: int,
        page_token: str
    ) -> Tuple[List[Relationship], ListResult]:
        """Retrieve relationships with pagination and optional filtering."""
        if not tenant_id:
            raise ValueError("tenant_id is required")

        opts = ListOptions(page_size=page_size, page_token=page_token)
        return await self.repo.list(tenant_id, source_node_id, target_node_id, rel_type, opts)
